### ISSUE TITLE
FACTION CAMP use manual time calculation instead of time_duration

### ISSUE THREAD

**Issue by cainiaowu** [CleverRaven/Cataclysm-DDA#24997]
**Describe the bug**  
This code cause some weird display overflow when the time to complete is out of original timescale.

**To Reproduce**  
1. Debug yourself some npc and mind control trait.
2. Build the base camp with debug spawn material until you reach recipe "faction_base_camp_6". 
3. see the "191 hours left".

**Expected behavior**  
It should be "7 days 23 hours left", also reinventing wheels.

**Screenshots**  
![image](https://user-images.githubusercontent.com/4143592/44329458-0f638f00-a497-11e8-9fc9-29c20df26b91.png)

**Versions and configuration(please complete the following information):**
master latest.

**Additional context**  
Faction camp code really needs a proper clean up, there are a lot of copy-paste code and the whole structure is quite byzantine.

**Comment by Break-Off** [comment]
Quite the word choice you've got there. I've been testing NPC camp lately as well, and have seen some inconsistencies and exploits:

1. Already constructed structures and furniture can be deconstructed and the resultant materials presented to the camp manager as new material for upgrades. For example, deconstruct crates and straw beds to get piles of straw, two by fours, nails...

2. Item sort feature (called Menial Labor in-game) is a nifty feature, but, as far as I've seen, sort points are not distributed when volume exceeds container capacity, basically overflowing to adjacent tiles.

3. Camp Manager is basically, code-wise, posted permanently as a guard inside the tent, so he can't be retrieved back as a follower and given orders. This is specially problematic if a monster tears down the tent and the manager goes out to fight it, because he can't be called back inside.

4. Former issue expands into a more problematic one: camp manager only accepts materials that are inside the tent AND in their line of sight, meaning that if the camp manager is outside fighting zeds and can't see the materials, he won't accept them. Working around this by presenting the material near they when outside doesn't work. Only workaround is tearing down the tent to let them see the materials.

Even worse, using sort function when leaving the materials outside sorts them into crates just outside camp manager's LoS, forcing the player to AIM the stuff into the tent, which is heavily time consuming when the materials are logs and such. Current workaround to this is changing sort points, but I don't know how the current code interprets player-made structures, overlapping or substituting them.

5. Food distributing is weird, since it applies penalties to the food's calories in function of rotting time. That means that some hard to make, high in nutrition, short shelf-life foods like meat pizza are applied a high penalty because it rots in 2 days, becoming useless as camp food. A sensible fix would be applying penalties to stacking foods of similar shelf-life. As in, you're trying to feed your camp 10 days of food that rots in 2 days, game should apply an incremental penalty in function of total calories versus rotting time.

6. Is possible that wander spawns don't take the camp into account, spawning zombies right in the middle of it if the wandering horde feature decides so. Haven't tested with fortifications yet, and so I don't know if from that moment is taken into account.

**Comment by cainiaowu** [comment]
Well, you may want to split them into various different issues so they get tracked better.

**Comment by mlangsdorf** [comment]
Working on upgrading your faction camp code!
0. Time strings have been replaced in the refactor.
1. Fixing this isn't trivial, but basically the long term plan will be to track what features/structures/terrain you actually have in your camp.  So if you deconstruct all the beds to craft new beds and end up with 2 beds, you won't have access to the missions that require 4 beds.
2. Menial sort is probably going to need to merged with Robik7's zone sort stuff.
3. Camp managers can be dismissed and rehired without penalty as of Monday, so that's mostly fixed.
4. I'm pretty sure the inventory function check's the player's line of sight, not the camp managers. I know from personal experience that I've gotten the camp manager to upgrade using supplies that were in the southmost chest by standing one tile NW of it and talking to the camp manager.
5. Your solution sounds really complicated.
6. Wander spawns are broken and will spawn willy-nilly on the edge of the reality bubble. I've proposed some solutions but Kevin doesn't like them (in fairness, they were bad ideas). Wander spawns spawning in the center of places they shouldn't be able to reach is a wander spawns problem, not a faction camp fortification problem.
